examples/SimpleERC20/Hydra.sol

```diff
@@ -78,131 +137,73 @@ contract SimpleERC20 {
+            newSig = bytes4(keccak256("transfer(address,uint256,address,uint256)"));
 
-        } else if ( sig == bytes4(sha3("transferFrom(address,address,uint256)")) ) {
-            newSig = bytes4(sha3("transferFrom(address,uint256,address,address,uint256)"));
+        } else if ( sig == bytes4(keccak256("transferFrom(address,address,uint256)")) ) {
+            newSig = bytes4(keccak256("transferFrom(address,uint256,address,address,uint256)"));
 
-        } else if ( sig == bytes4(sha3("approve(address,uint256)")) ) {
-            newSig = bytes4(sha3("approve(address,uint256,address,uint256)"));
+        } else if ( sig == bytes4(keccak256("approve(address,uint256)")) ) {
+            newSig = bytes4(keccak256("approve(address,uint256,address,uint256)"));
 
-        } else if ( sig == bytes4(sha3("allowance(address,address)")) ) {
-            newSig = bytes4(sha3("allowance(address,uint256,address,address)"));
+        } else if ( sig == bytes4(keccak256("allowance(address,address)")) ) {
+            newSig = bytes4(keccak256("allowance(address,uint256,address,address)"));
 
-        } else if ( sig == bytes4(sha3("deposit()")) ) {
-            newSig = bytes4(sha3("deposit(address,uint256)"));
+        } else if ( sig == bytes4(keccak256("deposit()")) ) {
+            newSig = bytes4(keccak256("deposit(address,uint256)"));
             // deposit has no output
             outputSize = 0x0;
 
-        } else if ( sig == bytes4(sha3("withdraw(uint256)")) ) {
-            newSig = bytes4(sha3("withdraw(address,uint256,uint256)"));
+        } else if ( sig == bytes4(keccak256("withdraw(uint256)")) ) {
+            newSig = bytes4(keccak256("withdraw(address,uint256,uint256)"));
 
         } else {
             revert();
         }
-        
-        // allocate memory to store the return value
-        uint256 retValMem;
-        uint256 maxOutputSize = MAX_HEAD_OUTPUT_SIZE;
-        assembly {
-            retValMem := mload(0x40)                    // free memory pointer
-            mstore(0x40, add(retValMem, maxOutputSize)) // allocate output size
-        }
-
-        uint256 x;
-        /* new call arguments are:
-         *  newSig       -> x[0x0..0x3]
-         *  msg.sender   -> x[0x4..0x23]
-         *  msg.value    -> x[0x24..0x43]
-         *  calldata[4:] -> x[0x44..]
-         *
-         * len(c) = 0x4 + 0x20 + 0x20 (calldatasize - 0x4)
-         *        = calldatasize + 0x40
-         */
-        assembly {
-            x := mload(0x40)                        // Get a free memory pointer
 
-            mstore(x, newSig)                       // the new call signature
-            mstore(add(x, 0x4), caller)             // the original sender
-            mstore(add(x, 0x24), callvalue)         // the call value
+        // add the sender and value to the call args
+        uint256 mem = _malloc(msg.data.length + 64);
+        _mstore(mem, bytes32(newSig));
```

Comment:
I checked that manually. You're right that we should have a check in the Hydra tests.

Reply:
Actually all the tests in `hydra/test/test_hydra.py` implicitly exercise this. I'm not sure how to best write a test that directly checks that the redirection inside the Meta Contract is correct.

Reply:
@ftramer One possibility is a calling contract that fails unless the data it's provided is exactly some value? 

Perhaps this will come up later in the security review / testing overview, or we can open an issue, but I don't think it's blocking for a merge as we don't really test this anywhere in the existing code.